readme.md
# font-carrier

<p>
<a href="https://www.npmjs.com/package/font-carrier"><img src="https://img.shields.io/npm/v/font-carrier.svg" alt="npm package"></a>
<a href="https://nodejs.org/en/about/releases/"><img src="https://img.shields.io/node/v/font-carrier.svg" alt="node compatibility"></a>
<a href="https://travis-ci.org/purplebamboo/font-carrier"><img src="https://travis-ci.org/purplebamboo/font-carrier.svg?branch=master" alt="build status"></a>
</p>

font-carrier 是一个功能强大的字体操作库，使用它你可以随心所欲地操作字体。

一个字体 font，包含若干字形 glyph。比如我们浏览器里渲染`我`，浏览器就会去当前设置的font里面找到 `我`对应的字形 glyph，使用它的形状来渲染。不同的字体的`我`的字形形状不一，所以才有差别。

font-carrier 封装了简单的 API，让你可以将某个 SVG，设置成一个字对应的字形。也可以通过解析已有字体，拿到某个字在这个字体下面对应的 SVG。让你通过 SVG 的维度随意修改字体展现样式。

我们不生产字体，我们只是字体的搬运工

- 需要图形操作客户端，请移步[panda](https://github.com/stormtea123/panda)
- 需要命令行解析代码，按需精简字体，请移步 [font-collector](https://github.com/JailBreakC/font-collector)

# Features

* 支持创建一个空白字体
* 支持解析已有字体(ttf，SVG)
* 支持使用 SVG 来设置字的展现
* 支持解析 SVG 的各种转换还有各种非 path 图形
* 支持针对某一个字，导出对应的 SVG
* 支持导出5种浏览器主流字体（ttf，eot，woff，woff2，svg）
* 支持设置各种字体相关内容

# Getting Start

如果对 iconfont 还不是很了解的，请先参考这篇[文章](http://purplebamboo.github.io/2014/01/09/iconfont/)

## Install

```sh
npm install font-carrier --save
```

## Usage

### Step-1：创建一个空白字体，或者解析一个已有的字体，这样都可以得到一个字体对象

```js
var fontCarrier = require('font-carrier')

//创建空白字体对象
var font = fontCarrier.create()

//从其他字体解析
var transFont = fontCarrier.transfer('./test/test.ttf')
```

### Step-2：拿到字体对象后，你就可以使用 SVG 随意操作字体了

```js
//可以设置某个字对应的形状,当然 unicode 也是支持的
font.setSvg('我', fs.readFileSync('./test/svgs/circle.svg').toString())

//也可以使用setGlyph可以设置更多信息
font.setGlyph('我', {
  glyphName: '我',
  horizAdvX: '1024',//设置这个字形的画布大小为1024
  svg:fs.readFileSync('./test/svgs/circle.svg').toString()
})

//可以针对字直接拿到对应的svg
var svg = font.getSvg('我')

//也可以先拿到对应的字形对象，再导出对应的svg
var glyph = transFont.getGlyph('我')
glyph.toSvg()
```

### Step-3：使用get,set各种操作完后，你可以选择导出字体

```js
// 默认会导出 svg，ttf，eot，woff，woff2 5种字体，
// 可以不传 path，这样会默认返回一个包含四个字体 buffer 的对象
font.output({
  path: './iconfont'
})
```

### Step-4：导出字体后就可以在 Web 中使用了

```html
<style type="text/css">
  @font-face {
    font-family: 'iconfont';
    src: url('iconfont.eot'); /* IE9 */
    src: url('iconfont.eot?#iefix') format('embedded-opentype'), /* IE6-IE8 */
    url('iconfont.woff') format('woff2'),
    url('iconfont.woff') format('woff'), /* chrome、firefox */
    url('iconfont.ttf') format('truetype'), /* chrome、firefox、opera、Safari, Android, iOS 4.2+*/
    url('iconfont.svg#iconfont') format('svg'); /* iOS 4.1- */
  }

  .iconfont {
    font-family: "iconfont";
    font-size: 16px;
    font-style: normal;
  }
</style>

<span class="iconfont">我</span>
//此时渲染出来的图形就是你设置的 SVG的样子
```

## Example

### 案例一

使用 SVG 设置一些 icon

```js
var fontCarrier = require('font-carrier')
// 创建空白字体，使用 SVG 生成字体
var font = fontCarrier.create()
var love = fs.readFileSync('./test/svgs/love.svg').toString()
var mail = fs.readFileSync('./test/svgs/mail.svg').toString()

//使用汉字
font.setGlyph('爱', {
  svg: love,
  glyphName: '爱'
})

//使用unicode
font.setSvg('&#xe601;', mail)

font.output({
  path: './test'
})
```

### 案例二

从其他字体导出一些图标到自己的库里

```js
var fontCarrier = require('font-carrier')
var transFont = fontCarrier.transfer('./test/test.ttf')//改成存在的字体文件地址
// 生成空白字体
var font = fontCarrier.create()

var gs = transFont.getGlyph('我是方正')
// 设置到空白字体里面
font.setGlyph(gs)

/ /这样 font 导出的字体里面就有了「我是方正」对应的 SVG 形状了
font.output({
  path: './test'
})

```

### 案例三

对中文字体精简

```js
var fontCarrier = require('font-carrier')
var transFont = fontCarrier.transfer('./test/test.ttf')
// 会自动根据当前的输入的文字过滤精简字体
transFont.min('我是精简后的字体，我可以重复')
transFont.output({
  path: './min'
})
```

## API

更多文档请看[这里](./doc/api.md)

## Test

先确保安装依赖包 `npm install` 再运行`npm test` 之后访问 `./test/index.html`

# [ChangeLog](changelog.md)

# [Licence](LICENSE)

MIT


changelog.md
# 0.3.1 (2021-05-11)
- 添加 .npmignore 文件，防止发布的包体积太大
- 更新 readme.md 中的徽标

# 0.3.0 (2021-05-11)
- `convert()` 和 `font.output()` 中的 `types` 选项将自动忽略不支持的格式
- 升级相关依赖
- 更新测试用例，生成 ttx 文件，方便查看升级后的字体变化
- 由于升级的 ttf2woff2 依赖 Node.js >=12.19.0，font-carrier 不再支持小于 12.19.0 的版本
# 0.2.1 (2021-03-16)
- Engine 中的 `convert()` 方法支持转换 ttf 文件（之前只支持 svg 字体）

# 0.2.0 (2019-09-06)
- New: 支持配置 svg2ttf 的 options 参数
- Fix: 部分字体基线错位
- Fix: 升级 svgpath，修复 [arcs 解析错误](https://github.com/fontello/svgpath/issues/23)
- Update: 升级除 ttf2svg 以外的所有依赖
- Doc: 增加 MIT LICENSE，增加 changelog.md


lib/helper/engine.js
var path = require('path')
var fs = require('fs');
var _ = require('lodash')
var helper = require('./helper.js')
var opentype = require('opentype.js')
var DOMParser = require('xmldom').DOMParser


var svg2ttf = require('svg2ttf')
var ttf2svg = require('ttf2svg')
var ttf2eot = require('ttf2eot')
var ttf2woff = require('ttf2woff')
var ttf2woff2 = require('ttf2woff2')

var fontEngine = {
  svg2svg: function(input) {
    if (Buffer.isBuffer(input)) return input
    return Buffer.from(input)
  },
  // TODO add test case
  svg2ttf: function(input, options) {
    if (Buffer.isBuffer(input)) input = input.toString()
    return Buffer.from(svg2ttf(input).buffer)
  },
  svg2eot: function(input, options) {
    var ttfBuffer = this.svg2ttf(input, options)
    return this.ttf2eot(ttfBuffer);
  },
  svg2woff: function(input, options) {
    var ttfBuffer = this.svg2ttf(input, options)
    return this.ttf2woff(ttfBuffer);
  },
  ttf2ttf: function(input) {
    if (Buffer.isBuffer(input)) return input
    return Buffer.from(input)
  },
  ttf2svg: function(input) {
    return Buffer.from(ttf2svg(input))
  },
  ttf2eot: function(input) {
    return Buffer.from(ttf2eot(input).buffer)
  },
  ttf2woff: function(input) {
    return Buffer.from(ttf2woff(input).buffer)
  },
  ttf2woff2: function(input) {
    return Buffer.from(ttf2woff2(input).buffer)
  },
}

var ALL_FONTS_TYPES = ['svg', 'ttf', 'eot', 'woff', 'woff2']



var _normalizeInput = function(input) {
  if (!input) return false
    //a buffer
  if (Buffer.isBuffer(input)) return input
    //a path
  if (_.isString(input) && fs.existsSync(input)) {
    return fs.readFileSync(input)
  }
  //a svg font string
  if (_.isString(input) && /font-face/.test(input)) {
    return Buffer.from(input)
  }
  return false
}

/**
 * 解析一个字体转换成js对象
 * @param  {[type]} input [description]
 * @return {[type]}         [description]
 */
fontEngine.parse = (function() {
  //参考  svg2ttf里面的解析代码
  //https://github.com/fontello/svg2ttf

  var _getGlyph = function(glyphElem) {
    var glyph = {};

    glyph.d = glyphElem.getAttribute('d').trim()
    glyph.unicode = helper.normalizeUnicode(glyphElem.getAttribute('unicode'))
    glyph.name = glyphElem.getAttribute('glyph-name')

    if (glyphElem.getAttribute('horiz-adv-x')) {
      glyph.horizAdvX = parseInt(glyphElem.getAttribute('horiz-adv-x'), 10)
    }

    return glyph;
  }

  var _parseTtfFont = function(ttfBuffer) {

    function toArrayBuffer(buffer) {
      var arrayBuffer = new ArrayBuffer(buffer.length)
      var data = new Uint8Array(arrayBuffer)
      for (var i = 0; i < buffer.length; i += 1) {
        data[i] = buffer[i]
      }

      return arrayBuffer
    }

    var font = opentype.parse(toArrayBuffer(ttfBuffer))

    if (!font.supported) {
      return helper.showError('Font is not supported (is this a Postscript font?)');
    }

    var hhea = font.tables.hhea
    var head = font.tables.head
    var name = font.tables.name
    var fontObjs = {
      options: {
        id: name.postScriptName.en || 'iconfont',
        horizAdvX: hhea.advanceWidthMax || 1024,
        vertAdvY: head.unitsPerEm || 1024
      },
      fontface: {
        fontFamily: name.fontFamily.en || 'iconfont',
        ascent: hhea.ascender,
        descent: hhea.descender,
        unitsPerEm: head.unitsPerEm
      },
      glyphs: {}
    }

    var path, unicode
    _.each(font.glyphs.glyphs, function(g) {
      try {
        path = g.path.toPathData()
        if (path && _.isArray(g.unicodes)) {
          _.each(g.unicodes,function(_unicode){
            unicode = '&#x' + (_unicode).toString(16) + ';'
            fontObjs.glyphs[unicode] = {
              d: path,
              unicode: unicode,
              name: g.name || 'uni' + _unicode,
              horizAdvX: g.advanceWidth,
              vertAdvY: fontObjs.options.vertAdvY
            }

          })
        }
      } catch (e) {
        //todo debug options
        //helper.showError(e)
      }

    })
    return fontObjs

  }

  var _parseSvgFont = function(svgBuffer) {
    var svgStr = svgBuffer.toString()
    var doc = (new DOMParser()).parseFromString(svgStr, 'application/xml')
    var fontElem = doc.getElementsByTagName('font')[0]
    var fontFaceElem = fontElem.getElementsByTagName('font-face')[0]

    var font = {
      options: {
        id: fontElem.getAttribute('id') || 'iconfont'
      },
      fontface: {
        fontFamily: fontFaceElem.getAttribute('font-family') || 'iconfont',
        fontStretch: fontFaceElem.getAttribute('font-stretch') || 'normal'
      },
      glyphs: {}
    }

    // Get <font-face> numeric attributes
    attrs = {
      ascent: 'ascent',
      descent: 'descent',
      unitsPerEm: 'units-per-em',
      fontWeight: 'font-weight'
    }

    _.forEach(attrs, function(val, key) {
      if (fontFaceElem.hasAttribute(val)) {
        font.fontface[key] = parseInt(fontFaceElem.getAttribute(val), 10)
      }
    })
    var fontfaceHeight = font.fontface.unitsPerEm


    var attrs = {
      horizAdvX: 'horiz-adv-x',
      vertAdvY: 'vert-adv-y'
        // horizOriginX: 'horiz-origin-x',
        // horizOriginY: 'horiz-origin-y',
        // vertOriginX: 'vert-origin-x',
        // vertOriginY: 'vert-origin-y'
    }
    _.forEach(attrs, function(val, key) {
      if (fontElem.hasAttribute(val)) {
        font.options[key] = parseInt(fontElem.getAttribute(val), 10)
      }
    })
    if (!font.options.vertAdvY) {
      font.options.vertAdvY = fontfaceHeight
    }
    if (!font.options.horizAdvX) {
      font.options.horizAdvX = fontfaceHeight
    }

    _.forEach(fontElem.getElementsByTagName('glyph'), function(glyphElem) {
      var glyph = _getGlyph(glyphElem)
      font.glyphs[glyph.unicode] = glyph

      if (!glyph.horizAdvX && font.options.horizAdvX) {
        glyph.horizAdvX = font.options.horizAdvX
      }
      if (!glyph.vertAdvY) {
        glyph.vertAdvY = font.options.vertAdvY ? font.options.vertAdvY : fontfaceHeight
      }

    })

    return font
  }

  var parse = function(input) {
    var inputBuffer = _normalizeInput(input)

    if (!inputBuffer) {
      helper.showError('font convert input error,support path,buffer,string！')
      return
    }

    if (/<svg/.test(inputBuffer.toString())) return _parseSvgFont(inputBuffer)

    //不是svg字体就默认是ttf字体,就使用ttf.js进行解析
    return _parseTtfFont(inputBuffer)
  }

  return parse

})()



/**
 * 一次性的生成其余几种字体,接受一个svg字体
 *
 * @param  {string|buffer|path} fontBuffer 字体输入可以是路径，buffer对象，string 对象
 * @param  {string} outputTypes 字体输出的类型，默认是所有四种，忽略其他不支持的格式
 * @param  {string} outputPath 字体的输出路径，不需要带后缀
 * @return {array}  outFonts    返回转换后的字体 buffer 数组
 */
fontEngine.convert = (function() {

  var _normalizePath = function(p) {
    if (!p) return false

    if (path.extname(p)) {
      p = p.replace(path.extname(p), '')
    }
    return p
  }

  var _convert = function(fontBuffer, outputTypes, options) {
    var result = {}
    var input = fontBuffer
    var ttfBuffer

    if (options.inputTypes === 'ttf') {
      ttfBuffer = fontEngine.ttf2ttf(input, options)
    } else {
      ttfBuffer = fontEngine.svg2ttf(input, options)
    }

    _.each(outputTypes, function(type) {
      if (type === 'svg') {
        result[type] = input
      } else {
        result[type] = fontEngine['ttf2' + type](ttfBuffer, options)
      }
    })

    return result
  }

  var _writeFile = function(output, outFonts) {
    _.each(outFonts, function(font, type) {
      fs.writeFileSync(output + '.' + type, font)
    })
  }

  return function(options) {
    var inputBuffer = _normalizeInput(options.input)
    var outputPath = _normalizePath(options.path)
    var outputTypes = options.types || ALL_FONTS_TYPES
    // 移除不支持的字体格式，防止报错
    outputTypes = outputTypes.filter(item => ALL_FONTS_TYPES.includes(item))

    if (!inputBuffer) {
      helper.showError('font convert input error,support path,buffer,string.please check if the file exist,while the input is a path.')
      return
    }

    var outFonts = _convert(inputBuffer, outputTypes, options)
    if (outputPath) {
      _writeFile(outputPath, outFonts)
    }
    return outFonts

  }
})()


/**
 * 字体转换引擎,提供 svg 到ttf, ttf 到 woff, woff2, eot, svg 的转换
 * 提供字体的解析与转换
 */
module.exports = fontEngine
